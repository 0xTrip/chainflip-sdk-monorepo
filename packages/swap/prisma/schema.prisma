// This is your Prisma schema file,
// learn more about it in the docs: https://pris.ly/d/prisma-schema

// `pnpm prisma format` after changes :)

generator client {
  provider        = "prisma-client-js"
  // https://github.com/prisma/prisma/issues/2584#issuecomment-646414549
  // https://github.com/prisma/prisma/issues/6603#issuecomment-849979849
  output          = "../node_modules/.prisma/client"
  previewFeatures = ["multiSchema"]
}

datasource db {
  provider = "postgresql"
  url      = env("DATABASE_URL")
  schemas  = ["public", "private"]
}

model SwapIntent {
  id             BigInt   @id @default(autoincrement())
  uuid           String   @unique @default(uuid())
  ingressAsset   String
  ingressAddress String
  egressAsset    String
  egressAddress  String
  active         Boolean  @default(true)
  createdAt      DateTime @default(now())
  swaps          Swap[]

  @@index([ingressAddress])
  @@schema("public")
}

model Swap {
  id                BigInt     @id @default(autoincrement())
  /// the swap id on the state chain
  nativeId          BigInt     @unique
  ingressAmount     Decimal    @db.Decimal(30, 0)
  ingressReceivedAt DateTime
  swapExecutedAt    DateTime?
  egressCompleteAt  DateTime?
  swapIntent        SwapIntent @relation(fields: [swapIntentId], references: [id])
  swapIntentId      BigInt
  egress            Egress?
  createdAt         DateTime   @default(now())
  updatedAt         DateTime   @default(now()) @updatedAt

  @@schema("public")
}

enum Network {
  Polkadot
  Ethereum
  Bitcoin

  @@schema("public")
}

model Egress {
  id          BigInt     @id @default(autoincrement())
  /// the egress id on the state chain
  nativeId    BigInt
  /// the swap id on the state chain
  network     Network
  amount      Decimal    @db.Decimal(30, 0)
  timestamp   DateTime
  swap        Swap?      @relation(fields: [swapId], references: [id])
  swapId      BigInt?    @unique
  broadcast   Broadcast? @relation(fields: [broadcastId], references: [id])
  broadcastId BigInt?
  createdAt   DateTime   @default(now())

  @@unique([nativeId, network])
  @@schema("public")
}

model Broadcast {
  id       BigInt   @id @default(autoincrement())
  /// the broadcast id on the state chain
  nativeId BigInt
  network  Network
  egresses Egress[]

  @@unique([nativeId, network])
  @@schema("public")
}

model State {
  id     Int @id @default(autoincrement())
  height Int @default(0)

  @@schema("private")
}

model MarketMaker {
  id        Int    @id @default(autoincrement())
  name      String @unique
  publicKey String

  @@schema("private")
}

enum ThirdPartyProtocol {
  lifi
  squid

  @@schema("public")
}

model ThirdPartySwap {
  id            BigInt             @id @default(autoincrement())
  uuid          String             @unique
  protocol      ThirdPartyProtocol
  routeResponse Json
  txHash        String
  txLink        String  
  createdAt     DateTime           @default(now())

  @@schema("public")
}
