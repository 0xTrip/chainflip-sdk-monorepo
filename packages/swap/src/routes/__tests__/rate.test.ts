// import BigNumber from 'bignumber.js';
import * as crypto from 'crypto';
import { once } from 'events';
import { Server } from 'http';
import { AddressInfo } from 'net';
import request from 'supertest';
import { promisify } from 'util';
import prisma from '../../client';
import QuotingClient from '../../quoting/QuotingClient';
import app from '../../server';
// import RpcClient from '../../utils/RpcClient';

const generateKeyPairAsync = promisify(crypto.generateKeyPair);

jest.mock(
  '../../utils/RpcClient',
  () =>
    class {
      async connect() {
        return this;
      }

      sendRequest() {
        throw new Error('unmocked request');
      }
    },
);

describe('server', () => {
  let server: Server;
  let client: QuotingClient;

  beforeEach(async () => {
    server = app.listen(0);
    await prisma.$queryRaw`TRUNCATE TABLE private."MarketMaker" CASCADE`;
    const name = 'web_team_whales';
    const pair = await generateKeyPairAsync('ed25519');
    await prisma.marketMaker.create({
      data: {
        name: 'web_team_whales',
        publicKey: pair.publicKey
          .export({ format: 'pem', type: 'spki' })
          .toString('base64'),
      },
    });

    client = new QuotingClient(
      `http://localhost:${(server.address() as AddressInfo).port}`,
      name,
      pair.privateKey.export({ format: 'pem', type: 'pkcs8' }).toString(),
    );
    await once(client, 'connected');
  });

  afterEach((cb) => {
    client.close();
    server.close(cb);
  });

  describe('GET /quote', () => {
    it('gets the rates', async () => {
      // TODO: restore broker mock
      // const sendSpy = jest
      //   .spyOn(RpcClient.prototype, 'sendRequest')
      //   .mockResolvedValueOnce(new BigNumber('1').times(10e18));
      const params = new URLSearchParams({
        ingressAsset: 'FLIP',
        egressAsset: 'ETH',
        amount: '1',
      });

      client.setQuoteRequestHandler(async (req) => ({
        id: req.id,
        intermediate_amount: '1000000000',
        egress_amount: '0.005',
      }));

      const { body, status } = await request(server).get(
        `/quote?${params.toString()}`,
      );

      expect(status).toBe(200);
      expect(body).toMatchSnapshot({ id: expect.any(String) });
      // expect(sendSpy).toHaveBeenCalledTimes(1);
    });
  });
});
