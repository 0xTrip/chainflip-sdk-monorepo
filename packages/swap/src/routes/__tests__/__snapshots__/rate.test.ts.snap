// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`server GET /quote gets the rates 1`] = `
{
  "egress_amount": "0.005",
  "id": Any<String>,
  "intermediate_amount": "1000000000",
}
`;
